examples/workflows/await_workflow.rb

```diff
@@ -0,0 +1,64 @@
+require 'activities/echo_activity'
+require 'activities/long_running_activity'
+
+# This example workflow exercises all three conditions that can change state that is being
+# awaited upon: activity completion, sleep completion, signal receieved.
+class AwaitWorkflow < Temporal::Workflow
+  def execute(expected_signal)
+    signals_received = {}
+
+    workflow.on_signal do |signal, input|
+      signals_received[signal] = input
+    end
+
+    workflow.await do
+      workflow.logger.info("Awaiting #{expected_signal}, signals received so far: #{signals_received}")
+      signals_received.key?(expected_signal)
+    end
+
+    # Run an activity but with a max time limit by starting a timer. This activity
+    # will not complete before the timer, which may result in a failed activity task after the
+    # workflow is completed.
+    long_running_future = LongRunningActivity.execute(15, 0.1)
+    timeout_timer = workflow.start_timer(1)
+    workflow.await do
+      long_running_future.finished? || timeout_timer.finished?
+    end
+
+    timer_beat_activity = timeout_timer.finished? && !long_running_future.finished?
+
+    activity_futures = {}
+    echos_completed = 0
+
+    10.times do |i|
+      workflow.await do
+        workflow.logger.info("Activities in flight #{activity_futures.length}")
+        activity_futures.length < 2
+      end
+
+      future = EchoActivity.execute("hi #{i}")
+      future.done do
+        activity_futures.delete(i)
```

Comment:
I wasn't sure what to do about this. It is a race condition in threaded code, but it's not based on how workflow code is currently executed. Ideas on how to make this more proper? I don't have much expertise with multi-threaded Ruby code.

Reply:
Just move the line `activity_futures[I] = future` to happen before you're setting the callback :) The reason it's not causing a race is because `EchoActivity.execute` doesn't start the activity straight away, but rather at the next blocking point

Reply:
Got it now. I think as written, it would fail if run in unit test mode, although it does in practice work fine running against a Temporal server.

Reply:
Yes, it works against the Temporal server because the commands are never dispatched before a blocking call (in this case it would be the next iteration's `.await`). But still a good idea to keep these things free from races